tests/unit/s2n_x509_validator_test.c

```diff
@@ -0,0 +1,619 @@
+    struct s2n_stuffer chain_in_stuffer, cert_stuffer, temp_out_stuffer;
+    s2n_stuffer_alloc_ro_from_string(&chain_in_stuffer, pem_data);
+    s2n_stuffer_growable_alloc(&cert_stuffer, 4096);
+    s2n_stuffer_growable_alloc(chain_out_stuffer, 4096);
+
+    uint32_t chain_size = 0;
+    do {
+        s2n_stuffer_certificate_from_pem(&chain_in_stuffer, &cert_stuffer);
+
+        uint32_t cert_len = s2n_stuffer_data_available(&cert_stuffer);
+        uint8_t *raw_cert_data = s2n_stuffer_raw_read(&cert_stuffer, cert_len);
+
+        if (cert_len) {
+            struct s2n_blob cert_data = {.data = raw_cert_data, .size = cert_len};
+            chain_size += cert_data.size + 3;
+            s2n_stuffer_write_uint24(chain_out_stuffer, cert_data.size);
+            s2n_stuffer_write(chain_out_stuffer, &cert_data);
+        }
+    } while (s2n_stuffer_data_available(&chain_in_stuffer));
+
+    s2n_stuffer_free(&cert_stuffer);
+    s2n_stuffer_free(&chain_in_stuffer);
+    return chain_size;
+}
+
+struct host_verify_data {
+    const char *name;
+    uint8_t found_name;
+    uint8_t callback_invoked;
+};
+
+static uint8_t verify_host_reject_everything(const char *host_name, size_t host_name_len, void *data) {
+    struct host_verify_data *verify_data = (struct host_verify_data *) data;
+    verify_data->callback_invoked = 1;
+    return 0;
+}
+
+static uint8_t verify_host_accept_everything(const char *host_name, size_t host_name_len, void *data) {
+    struct host_verify_data *verify_data = (struct host_verify_data *) data;
+    verify_data->callback_invoked = 1;
+    return 1;
+}
+
+
+static uint8_t verify_host_verify_alt(const char *host_name, size_t host_name_len, void *data) {
+    struct host_verify_data *verify_data = (struct host_verify_data *) data;
+
+    verify_data->callback_invoked = 1;
+    if (!strcmp(host_name, verify_data->name)) {
+        verify_data->found_name = 1;
+        return 1;
+    }
+
+    return 0;
+}
+
+int main(int argc, char **argv) {
+
+    BEGIN_TEST();
+
+    /* test empty trust store */
+    {
+        struct s2n_x509_trust_store trust_store;
+        s2n_x509_trust_store_init(&trust_store);
+
+        EXPECT_FALSE(s2n_x509_trust_store_has_certs(&trust_store));
+    }
+
+    /* test trust store from PEM file */
+    {
+        struct s2n_x509_trust_store trust_store;
+        s2n_x509_trust_store_init(&trust_store);
+        int err_code = s2n_x509_trust_store_from_ca_file(&trust_store, S2N_DEFAULT_TEST_CERT_CHAIN);
+        EXPECT_EQUAL(0, err_code);
+        EXPECT_TRUE(s2n_x509_trust_store_has_certs(&trust_store));
+        s2n_x509_trust_store_cleanup(&trust_store);
+    }
+
+    /* test trust store from non-existent PEM file */
+    {
+        struct s2n_x509_trust_store trust_store;
+        s2n_x509_trust_store_init(&trust_store);
+        int err_code = s2n_x509_trust_store_from_ca_file(&trust_store, "dskfjasdklfjsdkl");
```

Comment:
The keyboard mash is fine, but if you're interested; I think that '\0' and "/" are formally invalid filenames that are forbidden to exist on POSIX systems. That said, they may not really test what you want, I'm not sure what error code they return.